extras/indexing/src/main/java/org/apache/rya/indexing/mongodb/pcj/PcjQueryNode.java

```diff
@@ -0,0 +1,184 @@
+        final TupleExpr te = pq.getTupleExpr();
+        Preconditions.checkArgument(PCJOptimizerUtilities.isPCJValid(te), "TupleExpr is an invalid PCJ.");
+
+        final Optional<Projection> projection = new ParsedQueryUtil().findProjection(pq);
+        if (!projection.isPresent()) {
+            throw new MalformedQueryException("SPARQL query '" + sparql + "' does not contain a Projection.");
+        }
+        setProjectionExpr(projection.get());
+    }
+
+    /**
+     *
+     * @param accCon
+     *            - connection to a valid Accumulo instance
+     * @param tablename
+     *            - name of an existing PCJ table
+     * @throws PCJStorageException
+     * @throws MalformedQueryException
+     * @throws SailException
+     * @throws QueryEvaluationException
+     * @throws TableNotFoundException
+     * @throws AccumuloSecurityException
+     * @throws AccumuloException
+     */
+    public PcjQueryNode(final Configuration conf, final String tablename)
+            throws PCJStorageException, MalformedQueryException {
+        indexer = new MongoPrecomputedJoinIndexer();
+        pcjDocs = indexer.getPcjStorage(conf);
+        this.tablename = tablename;
+    }
+
+    /**
+     * returns size of table for query planning
+     */
+    @Override
+    public double cardinality() {
+        double cardinality = 0;
+        try {
+            cardinality = pcjDocs.getPcjMetadata(tablename).getCardinality();
+        } catch (final PcjException e) {
+            e.printStackTrace();
+        }
+        return cardinality;
+    }
+
+    @Override
+    public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(final BindingSet bindingset)
+            throws QueryEvaluationException {
+        return this.evaluate(Collections.singleton(bindingset));
+    }
+
+    /**
+     * Core evaluation method used during query evaluation - given a collection
+     * of binding set constraints, this method finds common binding labels
+     * between the constraints and table, uses those to build a prefix scan of
+     * the Accumulo table, and creates a solution binding set by iterating of
+     * the scan results.
+     *
+     * @param bindingset
+     *            - collection of {@link BindingSet}s to be joined with PCJ
+     * @return - CloseableIteration over joined results
+     */
+    @Override
+    public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(final Collection<BindingSet> bindingset)
+            throws QueryEvaluationException {
+
+        if (bindingset.isEmpty()) {
+            return new IteratorWrapper<BindingSet, QueryEvaluationException>(new HashSet<BindingSet>().iterator());
+        }
+        final CloseableIterator<BindingSet> iter = pcjDocs.getResults(tablename, new Authorizations(), bindingset);
+        return new CloseableIteration<BindingSet, QueryEvaluationException>() {
+            @Override
+            public boolean hasNext() throws QueryEvaluationException {
```

Comment:
You need to take more care with this iterator.  I notice that your getResults method for pcjDocs takes in the BindingSet as a constraint.  That's good.  You are taking into account the common Bindings between the constraint BindingSet and the docs in PcjDocStorage.  However, you are not joining the non-overlapping Bindings of the constraint BindingSets with the BindingSets returned by PcjDocumentStorage.  For example, if you have the BindingSet [a = <uri:Joe>, b = <uri:Bob>], and all BindingSets stored in PcjDocStorage have the form [a=..., c=...], you want to return BindingSets of the form [a = <uri:Joe>, b = <uri:Bob>, c =...].  That is, you need to include the b and c Bindings.  Where is this happening?